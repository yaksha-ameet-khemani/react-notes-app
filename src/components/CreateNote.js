import { useEffect, useState } from "react";
import axios from "axios";
import { useLocation, useHistory } from "react-router-dom";

const CreateNote = () => {

    const { state } = useLocation();
    const history = useHistory();
    const [note, setNote] = useState({
        title: "",
        description: "",
        author: "",
        status: false
    });

    useEffect(() => {
        if (state?.id) {
            axios.get("http://localhost:3500/notes/" + state?.id)
                .then(res => {
                    setNote({
                        title: res.data.title,
                        description: res.data.description,
                        author: res.data.author,
                        status: res.data.status
                    });
                })
                .catch(err => console.error(err));
        }
    }, [state?.id]);

    const changeHandler = (event) => {
        const name = event.target.name;
        const value = event.target.type == 'text' ? event.target.value : event.target.checked;
        setNote(note => {
            return {
                ...note,
                [name]: value
            }
        });
    }

    const submitHandler = (event) => {
        event.preventDefault();
        const id = state?.id ? state?.id : Math.floor((Math.random() * 999999999) + 111111111);
        const noteToCreate = {
            id: id,
            ...note
        }

        if (state?.id) {
            axios.put("http://localhost:3500/notes/" + state?.id, noteToCreate)
                .then(res => {
                    if (res.status == 200) {
                        history.push("/notes");
                    }
                })
                .catch(err => console.error(err));
        } else {
            axios.post("http://localhost:3500/notes", noteToCreate)
                .then(res => {
                    if (res.status == 201) {
                        history.push("/notes");
                    }
                })
                .catch(err => console.error(err));
        }
    }

    const cancelHandler = () => history.push("/");

    return <div>Create note
        <form>
            <label>Title</label>
            <input type="text" name="title" onChange={changeHandler} value={note.title} />
            <br />

            <label>Description</label>
            <input type="text" name="description" onChange={changeHandler} value={note.description} />
            <br />

            <label>Author</label>
            <input type="text" name="author" onChange={changeHandler} value={note.author} />
            <br />

            <label>Status</label>
            <input type="checkbox" name="status" value={note.status} checked={note.status} onChange={changeHandler} />
            <br />

            <button onClick={submitHandler}>{state?.id ? "Update" : "Submit"}</button>
            <button onClick={cancelHandler}>Cancel</button>
        </form>
    </div>
}

export default CreateNote;