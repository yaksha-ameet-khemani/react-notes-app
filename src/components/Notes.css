table {
    border: 1px solid black;
    border-collapse: collapse;
}

table tr {
    border: 1px solid black;
}

table td {
    border: 1px solid black;
}