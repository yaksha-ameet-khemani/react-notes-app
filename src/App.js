import { BrowserRouter, Link, Route, Switch } from "react-router-dom";
import CreateNote from "./components/CreateNote";
import Home from "./components/Home";
import Notes from "./components/Notes";

function App() {
  return (
    <>
      <div>My Notes</div>
      <BrowserRouter>
        <div>
          <Link to="/">Home</Link>----
          <Link to="/notes">All Notes</Link>----
          <Link to="/create-note">Create Note</Link>
        </div>
        <Switch>
          <Route exact path="/"><Home /></Route>
          <Route exact path="/notes"><Notes /></Route>
          <Route exact path="/create-note"><CreateNote /></Route>
        </Switch>
      </BrowserRouter>

    </>
  );
}

export default App;
