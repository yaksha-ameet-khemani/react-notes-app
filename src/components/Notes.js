import { useEffect, useState } from "react";
import axios from "axios";
import "./Notes.css";
import { useHistory } from "react-router-dom";

const Loading = () => <div>Loading...</div>

const Notes = () => {

    const [notes, setNotes] = useState({});
    const [copyNotes, setCopyNotes] = useState({});
    const history = useHistory();

    useEffect(() => {
        axios.get("http://localhost:3500/notes")
            .then(res => {
                const tempNotes = {};
                res.data.forEach(note => {
                    tempNotes[note.id] = {
                        title: note.title,
                        description: note.description,
                        author: note.author,
                        status: note.status
                    }
                });
                setNotes(tempNotes);
                setCopyNotes(res.data);
            })
            .catch(err => console.error(err));
    }, []);

    const createNoteHandler = () => history.push("create-note");

    const editHandler = id => history.push({
        pathname: "create-note",
        state: { id: id }
    });

    const deleteHandler = id => {
        axios.delete("http://localhost:3500/notes/" + id)
            .then(_ => {
                const tempNotes = Object.fromEntries(Object.entries(notes).filter(([key, _]) => key !== id));
                setNotes(tempNotes);
            })
            .catch(err => console.error(err));
    }

    const searchHandler = (e) => {
        const search = e.target.value;
        const toSearchNotes = copyNotes.map(note => {
            return {
                ...note,
                status: note.status ? "done" : "pending"
            }
        });
        const searchedNotes = toSearchNotes.filter(note => note.id.toString().includes(search) || note.author.includes(search) || note.status.includes(search));
        const tempNotes = {};
        searchedNotes.forEach(note => {
            tempNotes[note.id] = {
                title: note.title,
                description: note.description,
                author: note.author,
                status: note.status
            }
        });
        setNotes(tempNotes);
    }

    return <div>
        <button onClick={createNoteHandler}>Create Note</button>
        <input type="text" onChange={searchHandler} />
        <div>All Notes</div>
        {
            notes && Object.entries(notes).length ? <div>
                <table>
                    <thead>
                        <tr>
                            <th>S.No</th>
                            <th>ID</th>
                            <th>Title</th>
                            <th>Description</th>
                            <th>Author</th>
                            <th>Status</th>
                            <th></th>
                            <th></th>
                        </tr>
                    </thead>
                    <tbody>
                        {Object.entries(notes).map(([key, note], index) => <tr key={key}>
                            <td>{index + 1}</td>
                            <td>{key}</td>
                            <td>{note.title}</td>
                            <td>{note.description}</td>
                            <td>{note.author}</td>
                            <td>{note.status ? "Done" : "Pending"}</td>
                            <td><i className="material-icons" onClick={() => editHandler(key)}>edit</i></td>
                            <td><i className="material-icons" onClick={() => deleteHandler(key)}>delete</i></td>
                        </tr>)}
                    </tbody>
                </table>
            </div> : <Loading />
        }
    </div>
}

export default Notes;