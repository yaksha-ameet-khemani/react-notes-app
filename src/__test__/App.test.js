import React from 'react';
import { shallow } from 'enzyme';
import App from '../App';
import CreateNote from 'src/components/CreateNote';
import Notes from 'src/components/Notes';
import { render, screen } from '@testing-library/react'

// we need to mock react-router-dom for allowing useLocation hook in shallow copy
jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useLocation: () => ({
        pathname: "localhost:3000/create-note"
    })
}));

describe('boundary', () => {
    it(`AppTest boundary should mount App without crashing`, () => {
        const component = shallow(<App />);
        expect(component.getElements()).toMatchSnapshot();
        component.unmount();
    });

    it(`AppTest boundary should mount CreateNote without crashing`, () => {
        const component = shallow(<CreateNote />);
        expect(component.getElements()).toMatchSnapshot();
        component.unmount();
    });

    it(`AppTest boundary should mount Notes without crashing`, () => {
        const component = shallow(<Notes />);
        expect(component.getElements()).toMatchSnapshot();
        component.unmount();
    });
});

describe('functional', () => {
    it(`AppTest boundary must contain 3 Links`, () => {
        const component = shallow(<App />);
        expect(component.find(`Link`)).toHaveLength(3);
    });

    // it(`AppTest boundary links value`, () => {
    //     expect(screen.getByText('Home').closest('a')).toHaveAttribute('href', 'https://www.test.com/');
    // const linkEl = screen.getByRole('Link', { name: 'Home' });
    // expect(getByText("Click Me").href).toBe("https://www.test.com/")
    // expect(linkEl).toHaveAttribute('href', '...')
    // });
});
