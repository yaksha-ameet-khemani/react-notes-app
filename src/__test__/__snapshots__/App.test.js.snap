// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`boundary AppTest boundary should mount App without crashing 1`] = `
Array [
  <React.Fragment>
    <div>
      My Notes
    </div>
    <BrowserRouter>
      <div>
        <Link
          to="/"
        >
          Home
        </Link>
        ----
        <Link
          to="/notes"
        >
          All Notes
        </Link>
        ----
        <Link
          to="/create-note"
        >
          Create Note
        </Link>
      </div>
      <Switch>
        <Route
          exact={true}
          path="/"
        >
          <Home />
        </Route>
        <Route
          exact={true}
          path="/notes"
        >
          <Notes />
        </Route>
        <Route
          exact={true}
          path="/create-note"
        >
          <CreateNote />
        </Route>
      </Switch>
    </BrowserRouter>
  </React.Fragment>,
]
`;

exports[`boundary AppTest boundary should mount CreateNote without crashing 1`] = `
Array [
  <div>
    Create note
    <form>
      <label>
        Title
      </label>
      <input
        name="title"
        onChange={[Function]}
        type="text"
        value=""
      />
      <br />
      <label>
        Description
      </label>
      <input
        name="description"
        onChange={[Function]}
        type="text"
        value=""
      />
      <br />
      <label>
        Author
      </label>
      <input
        name="author"
        onChange={[Function]}
        type="text"
        value=""
      />
      <br />
      <label>
        Status
      </label>
      <input
        checked={false}
        name="status"
        onChange={[Function]}
        type="checkbox"
        value={false}
      />
      <br />
      <button
        onClick={[Function]}
      >
        Submit
      </button>
      <button
        onClick={[Function]}
      >
        Cancel
      </button>
    </form>
  </div>,
]
`;

exports[`boundary AppTest boundary should mount Notes without crashing 1`] = `
Array [
  <div>
    <button
      onClick={[Function]}
    >
      Create Note
    </button>
    <input
      onChange={[Function]}
      type="text"
    />
    <div>
      All Notes
    </div>
    <Loading />
  </div>,
]
`;
